src/us/kbase/test/auth2/service/AuthExternalConfigTest.java

```diff
@@ -0,0 +1,415 @@
+package us.kbase.test.auth2.service;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.fail;
+
+import java.net.URL;
+import java.util.Collections;
+import java.util.Map;
+
+import org.junit.Test;
+
+import com.google.common.collect.ImmutableMap;
+
+import nl.jqno.equalsverifier.EqualsVerifier;
+import us.kbase.auth2.lib.config.ConfigAction;
+import us.kbase.auth2.lib.config.ConfigAction.Action;
+import us.kbase.auth2.lib.config.ConfigAction.State;
+import us.kbase.auth2.lib.config.ConfigItem;
+import us.kbase.auth2.lib.exceptions.ExternalConfigMappingException;
+import us.kbase.auth2.lib.exceptions.IllegalParameterException;
+import us.kbase.auth2.service.AuthExternalConfig;
+import us.kbase.auth2.service.AuthExternalConfig.AuthExternalConfigMapper;
+import us.kbase.test.auth2.MapBuilder;
+import us.kbase.test.auth2.TestCommon;
+
+public class AuthExternalConfigTest {
+	
+	@Test
+	public void equals() {
+		EqualsVerifier.forClass(AuthExternalConfig.class).usingGetClass().verify();
+	}
+	
+	@Test
+	public void constructNoAction() throws Exception {
+		final AuthExternalConfig<Action> cfg = new AuthExternalConfig<>(
+				ConfigItem.noAction(),
+				ConfigItem.noAction(),
+				ConfigItem.noAction(),
+				ConfigItem.noAction(),
+				ConfigItem.noAction(),
+				ConfigItem.noAction());
+		
+		assertThat("incorrect trace", cfg.isIncludeStackTraceInResponse(),
+				is(ConfigItem.noAction()));
+		assertThat("incorrect def trace", cfg.isIncludeStackTraceInResponseOrDefault(), is(false));
+		assertThat("incorrect headers", cfg.isIgnoreIPHeaders(),
+				is(ConfigItem.noAction()));
+		assertThat("incorrect def headers", cfg.isIgnoreIPHeadersOrDefault(), is(false));
+		assertThat("incorrect log prefix", cfg.getAllowedLoginRedirectPrefix(),
+				is(ConfigItem.noAction()));
+		assertThat("incorrect link redirect", cfg.getCompleteLinkRedirect(),
+				is(ConfigItem.noAction()));
+		assertThat("incorrect login redirect", cfg.getCompleteLoginRedirect(),
+				is(ConfigItem.noAction()));
+		assertThat("incorrect post link redirect", cfg.getPostLinkRedirect(),
+				is(ConfigItem.noAction()));
+		
+		assertThat("incorrect toMap", cfg.toMap(), is(Collections.emptyMap()));
+	}
+	
+	@Test
+	public void constructRemove() throws Exception {
+		final AuthExternalConfig<Action> cfg = new AuthExternalConfig<>(
+				ConfigItem.remove(),
+				ConfigItem.remove(),
+				ConfigItem.remove(),
+				ConfigItem.remove(),
+				ConfigItem.remove(),
+				ConfigItem.remove());
```

Comment:
Is the idea behind passing `ConfigItem.remove()` that the default value for the config is unset? Not sure I follow what's being tested here.

Reply:
Similarly is `NoAction` just using a default value?

Reply:
This is part of a configuration system for the auth service that allows for the UI storing arbitrary configuration items in the auth database without having to change any other auth code (primarily the storage code and the core library code). I'm going to take advantage of that in a future PR to add more UI configuration without having to make any changes elsewhere.

The way it works is that the `AuthExternalConfig` class can either represent the state of the configuration (when the class is parameterized with `State`) or one or more actions to be taken to mutate the configuration state (when the class is parameterized with `Action`). So the same class is used to see the state or change the state.

Representing the state is pretty simple - there's either a value for each input (`ConfigItem.state(arg)`) or not (`ConfigItem.emptyState()). When representing a state mutation, instead of the state value an action is passed in:

`ConfigItem.noAction()` does what's on the tin.
`ConfigItem.remove()` removes the configuration item from the data storage system.
`ConfigItem.set(arg)` sets the configuration item to the new value.

When the `AuthExternalConfig` class (or any other class that implements `ExternalConfig`) gets passed into the storage system, the storage system calls the `toMap()` function to get a class-independent representation of the state mutation and applies it to the database. When getting the state, code passes in `AuthExternalConfigMapper` (or any other class that implements `ExternalConfigMapper`) which has a `fromMap()` method to take the class-independent representation of the state and create an `AuthExternalConfig` from it.